[Model Element] Attribute IBL resources to the WebContent process

achan00 opened the pull request:
#### ebb06fe8e3c04f6f7a657c22aa5dffd97991ba34
<pre>
[Model Element] Attribute IBL resources to the WebContent process
<a href="https://bugs.webkit.org/show_bug.cgi?id=293010">https://bugs.webkit.org/show_bug.cgi?id=293010</a>
<a href="https://rdar.apple.com/151255597">rdar://151255597</a>

Reviewed by Mike Wyrzykowski and Richard Robinson.

Get the textures from the EnvironmentResource asset and set their
memory attribution target to the WebContent process.

Guard WKSRKEntity with ENABLE(MODEL_PROCESS) as it&apos;s only used with
that enabled.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::setIBLAssetOwnership):
(WebKit::ModelProcessModelPlayerProxy::applyEnvironmentMapDataAndRelease):
(WebKit::ModelProcessModelPlayerProxy::applyDefaultIBL):
* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
(WKSRKEntity.applyIBL(_:attributionHandler:completion:)):
(WKSRKEntity.applyDefaultIBL(_:)):
(WKSRKEntity.applyIBL(_:completion:)): Deleted.
(WKSRKEntity.applyDefaultIBL): Deleted.
* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:
* Source/WebKit/WebKitSwift/StageMode/WKStageMode.h:
* Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift:

Canonical link: <a href="https://commits.webkit.org/295017@main">https://commits.webkit.org/295017@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b66db7bad0160182362a9fa945956a37476cd9bb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103683 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23385 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13705 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108873 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54335 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23727 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31931 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78783 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106689 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18400 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93531 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59117 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18210 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53711 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87993 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11638 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111262 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30839 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22737 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87776 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31200 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89732 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87426 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22286 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32305 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10033 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25196 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30767 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36070 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30561 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33896 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32122 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
```diff
@@ -313,7 +314,7 @@ public final class WKSRKEntity: NSObject {
     }
 
-    @objc(applyIBLData:withCompletion:) public func applyIBL(data: Data, completion: @escaping (Bool) -> Void) {
+    @objc(applyIBLData:attributionHandler:withCompletion:) public func applyIBL(data: Data, attributionHandler:@escaping (REAssetRef) -> Void, completion: @escaping (Bool) -> Void) {
```
```suggestion
    @objc(applyIBLData:attributionHandler:withCompletion:) public func applyIBL(data: Data, attributionHandler: @escaping (REAssetRef) -> Void, completion: @escaping (Bool) -> Void) {
```

rr-codes commented on Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
```diff
@@ -21,16 +21,16 @@
 // THE POSSIBILITY OF SUCH DAMAGE.
 
-#if os(visionOS)
+#if ENABLE_MODEL_PROCESS && os(visionOS)
```
the && os(visionOS) isn't needed anymore right?

achan00 replied:
Not right now because model process is only enabled for visionOS. I can remove it.

rr-codes commented on Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
```diff
@@ -347,13 +351,17 @@ public final class WKSRKEntity: NSObject {
 #if canImport(RealityFoundation, _version: 380)
                 let environment = await EnvironmentResource.defaultObject()
                 await MainActor.run {
+                    if let coreEnvironmentResourceAsset = environment.coreIBLAsset?.__as(REAssetRef.self) {
```
is there a non underscore version of `__as` too that you could use?

achan00 replied:
Unfortunately no.

rr-codes replied:
seeing as all this does is a forced cast anyways, maybe this can just be 

```suggestion
                    if let coreEnvironmentResourceAsset = environment.coreIBLAsset as? REAssetRef {
```

? I think this is better for multiple reasons:
1. It avoids using the weird `__as` function
2. It won't crash if it's not that type
3. It doesn't obfuscate what is actually happening (a cast)

achan00 replied:
Unfortunately that casts differently from `__as` and ends up returning null.

achan00 replied:
Checked with Richard offline and we'll stick with what we have for now.

rr-codes commented on Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:
```diff
@@ -24,11 +24,13 @@
+#import <wtf/Platform.h>
 
-#if defined(TARGET_OS_VISION) && TARGET_OS_VISION
+#if ENABLE(MODEL_PROCESS) && PLATFORM(VISION)
```
(ditto about the platform(vision) not being needed)

rr-codes commented on Source/WebKit/WebKitSwift/StageMode/WKStageMode.h:
```diff
@@ -23,9 +23,12 @@
 #import <Foundation/Foundation.h>
+#import <wtf/Platform.h>
+
+#if ENABLE(MODEL_PROCESS) && PLATFORM(VISION)
```
ditto

rr-codes commented on Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift:
```diff
@@ -21,7 +21,7 @@
 // THE POSSIBILITY OF SUCH DAMAGE.
 
-#if os(visionOS)
+#if ENABLE_MODEL_PROCESS && os(visionOS)
```
ditto

mwyrzykowski commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -818,28 +818,51 @@ static RESRT modelLocalizedTransformSRT(RESRT originalSRT)
     if (m_transientEnvironmentMapData) {
         if (m_transientEnvironmentMapData->size() > 0) {
-            [m_modelRKEntity applyIBLData:m_transientEnvironmentMapData->createNSData().get() withCompletion:makeBlockPtr([weakThis = WeakPtr { *this }] (BOOL succeeded) {
+            [m_modelRKEntity applyIBLData:m_transientEnvironmentMapData->createNSData().get() attributionHandler:makeBlockPtr([weakThis = WeakPtr { *this }] (REAssetRef coreEnvironmentResourceAsset) {
```
do you need to use `makeBlockPtr` here? Is it because this is non-ARC code?

In WebGPU we just say:
```
[objCObject someMethodTakingABlock:[weakTihs = WeakPtr { *this }]() {
        ...
}];
```

but WebGPU is ARC

achan00 replied:
This is non-ARC

mwyrzykowski reviewed (approved):
I defer to @rr-codes on the swift comments, otherwise 👍

rr-codes commented on Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
```diff
@@ -332,6 +332,10 @@ public final class WKSRKEntity: NSObject {
                 let environment = try await EnvironmentResource(cube: textureResource, options: .init())
 
                 await MainActor.run {
+                    if let coreEnvironmentResourceAsset = environment.coreIBLAsset?.__as(REAssetRef.self) {
```
ditto

rr-codes reviewed (approved):
approved (pending comments)

achan00 committed:
[Model Element] Attribute IBL resources to the WebContent process
https://bugs.webkit.org/show_bug.cgi?id=293010
rdar://151255597

Reviewed by Mike Wyrzykowski and Richard Robinson.

Get the textures from the EnvironmentResource asset and set their
memory attribution target to the WebContent process.

Guard WKSRKEntity with ENABLE(MODEL_PROCESS) as it's only used with
that enabled.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::setIBLAssetOwnership):
(WebKit::ModelProcessModelPlayerProxy::applyEnvironmentMapDataAndRelease):
(WebKit::ModelProcessModelPlayerProxy::applyDefaultIBL):
* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
(WKSRKEntity.applyIBL(_:attributionHandler:completion:)):
(WKSRKEntity.applyDefaultIBL(_:)):
(WKSRKEntity.applyIBL(_:completion:)): Deleted.
(WKSRKEntity.applyDefaultIBL): Deleted.
* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:
* Source/WebKit/WebKitSwift/StageMode/WKStageMode.h:
* Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift:

Canonical link: https://commits.webkit.org/295017@main
